src/js/media-tooltip.ts

```diff
@@ -0,0 +1,239 @@
+ * @cssproperty --media-font-size - `font-size` property.
+ * @cssproperty --media-text-content-height - `line-height` of button text.
+ *
+ * @cssproperty --media-tooltip-background - `background` color
+ * @cssproperty --media-tooltip-display - `display` of tooltip
+ * @cssproperty --media-tooltip-z-index - `z-index` of tooltip
+ * @cssproperty --media-tooltip-padding - `padding` of tooltip
+ * @cssproperty --media-tooltip-border-radius - `border-radius` of tooltip
+ * @cssproperty --media-tooltip-filter - `filter` property of tooltip, for drop-shadow
+ * @cssproperty --media-tooltip-white-space - `white-space` property of tooltip
+ * @cssproperty --media-tooltip-arrow-display - `display` property of tooltip arrow
+ */
+class MediaTooltip extends globalThis.HTMLElement {
+  static get observedAttributes(): string[] {
+    return [Attributes.POSITION, Attributes.CONTAINER];
+  }
+
+  arrowEl: HTMLElement;
+
+  constructor() {
+    super();
+
+    if (!this.shadowRoot) {
+      // Set up the Shadow DOM if not using Declarative Shadow DOM.
+      this.attachShadow({ mode: 'open' });
+      this.shadowRoot.appendChild(template.content.cloneNode(true));
+    }
+
+    this.arrowEl = this.shadowRoot.querySelector('#arrow');
+  }
+
+  // Adjusts tooltip position relative to the closest specified container
+  // such that it doesn't spill out of the left or right sides. Only applies
+  // to 'top' and 'bottom' positioned tooltips.
+  updateXOffset = () => {
+    const position = this.position;
+
+    // we don't offset against tooltips coming out of left and right sides
+    if (position === 'left' || position === 'right') {
+      // could have been offset before switching to a new position
+      this.style.removeProperty('--media-tooltip-offset-x');
+      return;
+    }
+
+    const containingSelector = this.container;
+    const containingEl = closestComposedNode(this, containingSelector);
+    if (!containingEl) return;
+    const { x: containerX, width: containerWidth } =
+      containingEl.getBoundingClientRect();
+    const { x: tooltipX, width: tooltipWidth } = this.getBoundingClientRect();
+    const tooltipRight = tooltipX + tooltipWidth;
+    const containerRight = containerX + containerWidth;
+    const offsetXVal = this.style.getPropertyValue('--media-tooltip-offset-x');
+    const currOffsetX = offsetXVal
+      ? parseFloat(offsetXVal.replace('px', ''))
+      : 0;
+
+    // we might have already offset the tooltip previously so we remove it's
+    // current offset from our calculations
+    const leftDiff = tooltipX - containerX + currOffsetX;
+    const rightDiff = tooltipRight - containerRight + currOffsetX;
+
+    // out of left bounds
+    if (leftDiff < 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${leftDiff}px`);
+      return;
+    }
+
+    // out of right bounds
+    if (rightDiff > 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${rightDiff}px`);
+      return;
+    }
```

Comment:
These can both be true at the same time, but it seems to be working ok. At a certain point the left edge becomes dominant, which seems fine. Or at least I can't think of how I'd want it to work differently.

![image](https://github.com/user-attachments/assets/6280730a-bd61-41b9-9077-2104870a3761)
![image](https://github.com/user-attachments/assets/a2137aa9-4557-45fc-adc3-75d16aab2fa2)

Reply:
Yeah I gave a little though to this but couldn't come up with a clean "this is how it _should_ look" in these types of scenarios and left it currently as undefined behaviour or a design issue instead of an implementation issue. We could try and detect this and fall back to centering? but it would still be visually strange.

Reply:
Yeah, works for now. 👍  If there's real issues with this we could look into a max-width and line wrapping.